Fix ProductionAirdrop aircraft not getting removed instantly

reaperrr opened the pull request:
...when leaving map before finishing TakeOff.

Fixes #16824, as far as I can tell.

reaperrr committed:
Fix ProductionAirdrop aircraft removal timing

...when leaving map before finishing TakeOff.

pchote commented on OpenRA.Mods.Cnc/Traits/Buildings/ProductionAirdrop.cs:
```diff
@@ -83,7 +83,13 @@ public override bool Produce(Actor self, ActorInfo producee, string productionTy
+				if (aircraftInfo.VTOL)
+					actor.QueueActivity(new Fly(actor, Target.FromCell(w, endPos)));
+				else
+					actor.QueueActivity(new FlyOffMap(actor));
```
Why can't VTOLs use `FlyOffMap`?

reaperrr replied:
Because `FlyOffMap` just directly ticks `FlyTick`, meaning it will ignore the `VTOL` boolean and take off diagonally.
And we can't use child activities in `FlyOffMap` (at least not for horizontal flying), or we lose the ability to exit the activity as soon as we leave the map.
Though I guess I could queue `TakeOff` as child activity there if it's a VTOL.

reaperrr commented:
Updated. Tested both TD and D2k, both still seem to work as intended.

pchote commented on OpenRA.Mods.Common/Activities/Air/FlyOffMap.cs:
```diff
@@ -35,6 +35,14 @@ public override bool Tick(Actor self)
+				return false;
+			}
+
+			// Using FlyTick directly is necessary to be able to exit the activity as soon as we leave the map
```
Actually not true (!)

Activities can set `ChildHasPriority = false` in the constructor to turn off the logic where their children tick before they do. We could queue the `TakeOff` and `Fly` from `OnFirstRun` and simplify `Tick` down to the outside map check followed by `TickChild(self);`

fruestueck reviewed (commented):
Works as intended. Can't comment the code.

reaperrr commented:
Updated again.

I intentionally made 2 ctors, one with target and one without, as the former will come in handy if we ever want aircraft to leave in a specific direction instead of just flying forward.

tovl commented on OpenRA.Mods.Common/Activities/Air/FlyOffMap.cs:
```diff
@@ -11,16 +11,43 @@
+			if (aircraft.Info.VTOL && self.World.Map.DistanceAboveTerrain(aircraft.CenterPosition) != aircraft.Info.CruiseAltitude)
+				QueueChild(new TakeOff(self));
+
+			QueueChild(new FlyForward(aircraft));
```
Instead of creating a new activity for this, couldn't we just call `FlyTimed` here with `ticks = -1`?

reaperrr replied:
Good idea.

reaperrr commented:
Updated.

pchote commented:
The introduced a small visual regression by removing the plane as soon as it hits the map edge - meaning the back half of the aircraft and all of its contrail blink out of existence. Can we please queue another few-cell times fly between the `FlyOffMap` and `RemoveSelf`?

abcdefg30 commented:
Shouldn't Fly*Off*Map already do this itself? (Not stop at the map edge but go a bit further?)

reaperrr commented:
> Shouldn't FlyOffMap already do this itself? (Not stop at the map edge but go a bit further?)

Well, technically it _does_ move them off map (but only by 1 tick). 
But I agree that `FlyOffMap` should handle this itself, else we'd have to insert that additional `FlyTimed` everywhere (airstrikes, crate planes and so on).

Updated.

pchote commented on OpenRA.Mods.Common/Activities/Air/FlyOffMap.cs:
```diff
@@ -11,32 +11,64 @@
+				leftMap = true;
+
+			// Wait a few ticks so the aircraft is completely off-map before ending activity
+			if (leftMap && offMapTicks++ > 25)
```
IMO it would be good practice to expose the 25 as an argument (which can still default to 25) in the constructor.

reaperrr replied:
OK.
But for the record, I tested higher values and for whatever crazy reasons they made the TD C17 circle back into the map before getting removed. And yes, I even disabled the 'nudge aircraft back into map' logic.

I'm too fed up with the aircraft code to debug that further for the forseeable future, though.

reaperrr commented:
Updated.

reaperrr committed:
Don't exit FlyOffMap immediately on leaving map

Usually they'll get removed afterwards, so they need to be
out of players' sight before ending this activity.

pchote commented:
This wasn't quite right, so I tweaked and pushed a fix. This fixes aircraft still vanishing immediately if the aircraft was already at cruise altitude, and I think it also fixes your circling issue.

pchote reviewed (dismissed):
:+1: with my tweaks

abcdefg30 commented on OpenRA.Mods.Common/Activities/Air/FlyOffMap.cs:
```diff
@@ -11,32 +11,62 @@
+
+		public FlyOffMap(Actor self, Target target, int endingDelay = 25)
+		{
+			aircraft = self.Trait<Aircraft>();
```
Imho this ctor should call the other instead of duplicating things.

reaperrr commented:
Updated.

abcdefg30 commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/666b2a9b2dedcb0240e1cc04d7b6e4deea88090a...c7954b24ac145ec20a21f61bf775cffd591fe85a)
